packages/contracts/contracts/universal/MultiSigWallet.sol

```diff
@@ -0,0 +1,529 @@
+    */
+    function _changeNumConfirmationRequirement(uint _numConfirmationsRequired)
+        internal
+        onlyWallet
+        validNumConfirmations(owners.length, _numConfirmationsRequired)
+    {
+        numConfirmationsRequired = _numConfirmationsRequired;
+    }
+
+    /**
+    * @notice Allows to change the number of required confirmations. Transaction has to be sent by wallet.
+    *
+    * @param _numConfirmationsRequired Number of required confirmations.
+    */
+    function changeRequirement(uint _numConfirmationsRequired)
+        external
+        override
+    {
+        _changeNumConfirmationRequirement(_numConfirmationsRequired);
+        emit RequirementChange(_numConfirmationsRequired);
+    }
+
+    /**
+    * @notice Allows an owner to submit and confirm a transaction.
+    *
+    * @param _destination Transaction target address.
+    * @param _value Transaction ether value.
+    * @param _data Transaction data payload.
+    * @return Returns transaction ID.
+    */
+    function submitTransaction(address _destination, uint _value, bytes memory _data)
+        public
+        override
+        virtual
+        returns (uint)
+    {
+        uint transactionId = addTransaction(_destination, _value, _data);
+        confirmTransaction(transactionId);
+        return transactionId;
+    }
+
+    /**
+    * @notice Allows an owner to confirm a transaction.
+    *
+    * @param _transactionId Transaction ID.
+    */
+    function confirmTransaction(uint _transactionId)
+        public
+        override
+        virtual
+        ownerExists(msg.sender)
+        transactionExists(_transactionId)
+        notConfirmed(_transactionId, msg.sender)
+    {
+        confirmations[_transactionId][msg.sender] = true;
+        emit Confirmation(msg.sender, _transactionId);
+        _executeTransaction(_transactionId);
+    }
+
+    /**
+    * @notice Allows an owner to revoke a confirmation for a transaction.
+    *
+    * @param _transactionId Transaction ID.
+    */
+    function revokeConfirmation(uint _transactionId)
+        external
+        override
+        virtual
+        ownerExists(msg.sender)
+        confirmed(_transactionId, msg.sender)
+        notExecuted(_transactionId)
+    {
+        confirmations[_transactionId][msg.sender] = false;
+        emit Revocation(msg.sender, _transactionId);
+    }
+
+    /**
+    * @notice Internal functions. Execute a confirmed transaction.
+    *
+    * @param _transactionId Transaction ID.
+    */
+    function _executeTransaction(uint _transactionId)
+        internal
+        ownerExists(msg.sender)
+        confirmed(_transactionId, msg.sender)
+        notExecuted(_transactionId)
+    {
+        if (_isConfirmed(_transactionId)) {
+            Transaction storage txn = transactions[_transactionId];
+            txn.executed = true;
+            (bool success, ) = txn.destination.call(
+                txn.data
+            );
```

Comment:
Call should be treated with a care. Please add a `nonReentant` or call with `SafeCall` variants.

Reply:
NonReentrant has been applied.